flesh out teaching and enumerate audiences (#14)
rgommers/peps: pep-9999.rst

Reviewer: You might want to add yourself to the author list, @msarahan :)

How to Teach This
=================

> There are at least three audiences that need to learn a workflow here.

Reviewer: ```suggestion
There are at least four audiences that need to learn a workflow here.
```

> 1. Python package maintainers wishing to express an external dependency
> 2. Package ecosystem maintainers, who are responsible for keeping the
>    mapping for their ecosystem current.
> 3. Core registry maintainers, who are responsible for curating the central
>    repository of PURL identifiers and descriptors
> 4. End users of packages that have external dependency metadata

Reviewer: ```suggestion
1. Python package maintainers wishing to express an external dependency.
2. Package ecosystem maintainers, who are responsible for keeping the
   mapping for their ecosystem current.
3. Core registry maintainers, who are responsible for curating the central
   repository of PURL identifiers and descriptors.
4. End users of packages that have external dependency metadata.
```

For consistency with the other lists.

Python package maintainer usage
-------------------------------

A package maintainer's responsibility is to decide the PURL that best
represents the external dependency that their package needs. Their task
consists of:

1. Understanding the nature of their dependency. Do they only need runtime 

>    libraries, or do they need development packages for build-time concerns.

Reviewer: ```suggestion
   libraries, or do they need development packages for build-time concerns?
```

   This understanding feeds into PEP 725, which specifies the expression of
   these dependencies in metadata.
2. Looking up the PURL. This can either mean knowing the name of the package 
   in their package ecosystem, and then inverse-mapping that to the PURL, or
   it can mean looking up the PURL directly.
3. When a package maintainer does not find an appropriate mapping, they should look 

>    for the PURL. It can be the case that although a PURL is registered, not every 

Reviewer: ```suggestion
   for a fitting PURL. It can be the case that although a PURL is registered, not every 
```

   the package maintainer may consider submitting a new PURL to the central registry.

An overall workflow diagram might look like this:

.. mermaid::

   flowchart TD
      A[Python package author with new external dependency] --> |Looks in| B(PURL/description collection)
      B --> | Find PURL OK | E(Add PURL to pyproject.toml)
      A --> | Looks in | C(Ecosystem mapping file)
      C --> | Finds familiar ecosystem package name | D(Inverse map ecosystem package name to PURL)
      D --> | Mapping exists | E
      B --> | PURL not found | F(Submit PURL proposal to PURL/description collection)
      F --> | Accepted | G(Mapping maintainers notified of missing PURL mappings)
      D --> | Mapping missing. User looks in PURL collection. | B
      B --> | Was mapping missing? | H(User may contribute entry to mapping)

Package ecosystem maintainers usage
-----------------------------------

Any packages that express a PURL dependency that does not have a mapping in a given package
ecosystem will mean a failed installation. It is thus critically important that each package
ecosystem maintain their mappings. Key to this will be automation. Some ideas for automation
are:

> 1. Alert mapping maintainers whenever a new PURL is added to the registry (probably noisy)
> 2. Provide tools that allow maintainers to diff their mappings to the registry contents to
>    quickly identify missing entries
> 3. Provide automated tooling that submits PRs to known mapping locations, such that maintainers
>    need only fill in the ecosystem package name
> 4. Provide status for each PURL, to readily identify which PURLs need attention

Reviewer: ```suggestion
1. Alert mapping maintainers whenever a new PURL is added to the registry (probably noisy).
2. Provide tools that allow maintainers to diff their mappings to the registry contents to
   quickly identify missing entries.
3. Provide automated tooling that submits PRs to known mapping locations, such that maintainers
   need only fill in the ecosystem package name.
4. Provide status for each PURL, to readily identify which PURLs need attention.
```

This maintenance is likely to be a lot of work to establish the initial mapping, but ideally small
on an ongoing basis.

Central PURL registry maintainers
---------------------------------

Central PURL registry maintainers curate the collection of PURLs. These contributors
need to be able to refer to clearly defined rules for when a new PURL can be defined. It is
undesirable to be loose with PURL definition, because each definition implies maintenance in
the mappings in many other places.

End user package consumers
--------------------------

In an ideal case, these package install commands can be done transparently, and the user 
experience is the same as the self-contained binary wheels of today. There are several
foreseeable issues that will arise, though:

> * A mapping does not exist for the user's desired package ecosystem
> * A user does not have permissions to run the install commands provided by our 
>   tool (e.g. system python users)

Reviewer: ```suggestion
* A mapping does not exist for the user's desired package ecosystem.
* A user does not have permissions to run the install commands provided by our 
  tool (e.g. system python users).
```

If packagers suddenly drop self-contained wheels before the mapping can be well-established,
there will likely be a period of negative user experience.

> TODO: can we set a target adoption level before we think it's safe to switch?

Reviewer: How do we measure this?

Author: I was thinking fractional download counts. Like for a given project, how many of the downloads of the latest version are the self-contained wheels vs. the shared-library variety?

Reviewer: If we can track both separately, that works, yes.

> If packagers suddenly drop self-contained wheels before the mapping can be well-established,
> there will likely be a period of negative user experience.
> 
> TODO: can we set a target adoption level before we think it's safe to switch?

Author: ```suggestion
If packagers suddenly drop self-contained wheels before the mapping can be well-established,
there will likely be a period of negative user experience. Packages should individually decide for themselves when it is "safe" to stop providing standalone wheels. One metric that might serve this purpose would be fractional download counts. For example, how many of the downloads of the latest version of a project are the self-contained wheels vs. the shared-library variety? This is likely going to be different on each platform.

TODO: can we set a target adoption level before we think it's safe to switch?
```

Reference Implementation
========================

A reference implementation should include three components:

1. A central registry that captures at a minimum PURL and description. This registry should 
   NOT contain specifics of package ecosystem mappings.

> 2. A standard specification for a collection of mappings (TODO: JSONSchema?)

Author: ```suggestion
2. A standard specification for a collection of mappings. JSONSchema is widely used for schema in many text editors, and would be a natural choice for expression of the standard specification.
```

> 2. A standard specification for a collection of mappings (TODO: JSONSchema?)

Reviewer: Sounds good!

3. An implementation of (2), providing mappings from the contents of the central 

>    registry to the ecosystem-specific package names

Reviewer: ```suggestion
   registry to the ecosystem-specific package names.
```

> TODO: link to proof-of-concept implementation for Grayskull.

Reviewer: ```suggestion
A prototype proof of concept implementation was contributed to Grayskull, a conda recipe generator for Python packages, via `conda/grayskull#518 <https://github.com/conda/grayskull/pull/518>`__.
```

ecosystem will mean a failed installation. It is thus critically important that each package
ecosystem maintain their mappings. Key to this will be automation. Some ideas for automation
are:

1. Alert mapping maintainers whenever a new PURL is added to the registry (probably noisy).
2. Provide tools that allow maintainers to diff their mappings to the registry contents to
   quickly identify missing entries.
3. Provide automated tooling that submits PRs to known mapping locations, such that maintainers
   need only fill in the ecosystem package name.
4. Provide status for each PURL, to readily identify which PURLs need attention.

This maintenance is likely to be a lot of work to establish the initial mapping, but ideally small
on an ongoing basis.

Central PURL registry maintainers
---------------------------------

Central PURL registry maintainers curate the collection of PURLs. These contributors
need to be able to refer to clearly defined rules for when a new PURL can be defined. It is
undesirable to be loose with PURL definition, because each definition implies maintenance in
the mappings in many other places.

> End user package consumers

Reviewer 2: I'd aim for: "no change in user experience by default, and end users do not need to know about this mechanism unless they opt in (which they may want to do to, for example, reduce their bandwidth and disk space usage".
